closure/private/defs.bzl

```diff
@@ -61,6 +61,84 @@ CLOSURE_JS_TOOLCHAIN_ATTRS = {
     "_unusable_type_definition": UNUSABLE_TYPE_DEFINITION,
 }
 
+ClosureJsLibraryInfo = provider("ClosureJsLibraryInfo", fields = {
+    "info": """
+File pointing to a ClosureJsLibrary protobuf file in pbtxt format
+that's generated by this specific Target. It contains some metadata
+as well as information extracted from inside the srcs files, e.g.
+goog.provide'd namespaces. It is used for strict dependency
+checking, a.k.a. layering checks.
+""",
+    "infos": """
+NestedSet<File> of all info files in the transitive closure. This
+is used by JsCompiler to apply error suppression on a file-by-file
+basis.""",
+    "ijs": "",
+    "ijs_files": "",
+    "srcs": """
+NestedSet<File> of all JavaScript source File artifacts in the
+transitive closure. These files MUST be JavaScript.""",
+    "js_module_roots": """
+NestedSet<String> of all execroot path prefixes in the transitive
+closure. For very simple projects, it will be empty. It is useful
+for getting rid of Bazel generated directories, workspace names,
+etc. out of module paths.  It contains the cartesian product of
+generated roots, external repository roots, and includes
+prefixes. This is passed to JSCompiler via the --js_module_root
+flag. See find_js_module_roots() in defs.bzl.            """,
+    "modules": """
+NestedSet<String> of all ES6 module name strings in the transitive
+closure. These are generated from the source file path relative to
+the longest matching root prefix. It is used to guarantee that
+within any given transitive closure, no namespace collisions
+exist. These MUST NOT begin with "/" or ".", or contain "..".""",
+    "descriptors": """
+NestedSet<File> of all protobuf definitions in the transitive
+closure. It is used so Closure Templates can have information about
+the structure of protobufs so they can be easily rendered in .soy
+files with type safety. See closure_js_template_library.bzl.""",
+    "stylesheets": """
+NestedSet<Label> of all closure_css_library rules in the transitive
+closure. This is used by closure_js_binary can guarantee the
+completeness of goog.getCssName() substitutions.""",
+    "has_closure_library": """
+Boolean indicating indicating if Closure Library's base.js is part
+of the srcs subprovider. This field exists for optimization.""",
+    "language": "",
+})
+
+ClosureExportsInfo = provider("ClosureExportsInfo", fields = {"exports": """
```

Comment:
Since we are introducing proper providers, would it be possible to get rid of this as well?

Reply:
For `ClosureExportsInfo`, it looks possible to remove it, by adding `exports` to `ClosuresJsLibraryInfo` and `WebFilesInfo`.

`ClosureJsLegacyRunfilesInfo` looks easier to remove, by using regular runfiles instead.

Both of those removals require some effort and should probably be done by code owners and not as a part of a large cleanup, where we're fixing about 80 other rulesets.

Reply:
Ok that's fair.